phi/torch/cuda/src/phi_torch_cuda.cpp

```diff
@@ -0,0 +1,337 @@
+        return res;
+    }
+
+    template<typename T>
+    void print_variable(char* name, T variable) {
+        std::cout << name << "\n" << variable << std::endl;
+    }
+
+    /*
+        Sparse versus Dense Matrix Multiplication. 32 floating point precision.
+        This function is called from cusparse_SpMM()
+    */
+    torch::Tensor floatSpMM(const at::Tensor& dA_csrOffsets,
+                            const at::Tensor& dA_columns,
+                            const at::Tensor& dA_values,
+                            at::Tensor& dB,
+                            const int64_t dim_i,
+                            const int64_t dim_j)
+    {
+        cusparseHandle_t     handle = NULL;
+        cusparseSpMatDescr_t matA;
+        cusparseDnVecDescr_t vecX, vecY;
+        auto A_nnz          = dA_values.size(0);
+        void*                dBuffer    = NULL;
+        size_t               bufferSize = 0;
+        float alpha           = 1.0f;
+        float beta            = 0.0f;
+
+        dB = dB.view({dim_j});
+        CHECK_CUSPARSE( cusparseCreate(&handle) )
+        // Create sparse matrix A in CSR format
+        CHECK_CUSPARSE( cusparseCreateCsr(&matA, dim_i, dim_j, A_nnz,
+                                          dA_csrOffsets.data_ptr<int64_t>(), dA_columns.data_ptr<int64_t>(),
+                                          dA_values.data_ptr<float>(),
+                                          CUSPARSE_INDEX_64I, CUSPARSE_INDEX_64I,
+                                          CUSPARSE_INDEX_BASE_ZERO, CUDA_R_32F) )
+        // Create dense vector X
+        CHECK_CUSPARSE( cusparseCreateDnVec(&vecX, dim_j, dB.data_ptr<float>(), CUDA_R_32F) )
+        // Create dense vector y
+        torch::Tensor dC = torch::empty({dim_i}, dB.options());
+        CHECK_CUSPARSE( cusparseCreateDnVec(&vecY, dim_i, dC.data_ptr<float>(), CUDA_R_32F) )
+        // allocate an external buffer if needed
+        CHECK_CUSPARSE( cusparseSpMV_bufferSize(
+                                     handle, CUSPARSE_OPERATION_NON_TRANSPOSE,
+                                     &alpha, matA, vecX, &beta, vecY, CUDA_R_32F,
+                                     CUSPARSE_CSRMV_ALG2, &bufferSize) )
+        CHECK_CUDA( cudaMalloc(&dBuffer, bufferSize) )
+
+        // execute SpMV
+        CHECK_CUSPARSE( cusparseSpMV(handle, CUSPARSE_OPERATION_NON_TRANSPOSE,
+                                     &alpha, matA, vecX, &beta, vecY, CUDA_R_32F,
+                                     CUSPARSE_CSRMV_ALG2, dBuffer) )
+
+        // destroy matrix/vector descriptors
+        CHECK_CUSPARSE( cusparseDestroySpMat(matA) )
+        CHECK_CUSPARSE( cusparseDestroyDnVec(vecX) )
+        CHECK_CUSPARSE( cusparseDestroyDnVec(vecY) )
+        CHECK_CUSPARSE( cusparseDestroy(handle) )
+        dB = dB.view({dim_j, 1});
+        dC = dC.view({dim_i, 1});
+        return dC;
+    }
+
+    /*
+        Sparse versus Dense Matrix Multiplication. 64 floating point precision.
```

Comment:
Same as above.

Reply:
/*
        - "Sparse CSR * Dense" matrix multiplication.
        - 64 bit floating point precision.
        - This function is called from cusparse_SpMM()

        The algorithm that is chosen to perform the multiplication is the following:
        CUSPARSE_SPMV_CSR_ALG2. Reference: https://docs.nvidia.com/cuda/cusparse/index.html

        This algorithm has been chosen because:
        1. Why Matrix*Vector operation?: The sparse * dense matrix multiplications that are performed in the solution of the
        conjugate gradient method are always [Sparse matrix * Dense vector]. A Matrix * Matrix multiplication could
        also be used. This would imply making sure that we are interpreting the dense matrix with the right dimensions
        (PyTorch tensors are row-major. CUDA is column-major). On the other hand a Vector is dimension agnostic and
        this requires no extra work.
        2. Why Algorithm 2?: The algorithm 1 is faster, but it does not provide deterministic results. Which may lead
        to a non-convergent solution.
        3. Why CUSPARSE_OPERATION_NON_TRANSPOSE?: The routine is 3x faster when we do not transpose the matrix.
    */